Dockerfile

```
@@ -1,29 +1,62 @@
 # Stage 1: Build golang dependencies and binaries
-FROM ubuntu:25.04 AS build
+FROM golang:1.24-bullseye AS build
 
 ARG VM_DEBUG
-
-
-RUN apt-get -qq update && \
-    apt-get -qq install curl build-essential git golang upx-ucl libjemalloc-dev libjemalloc2 libbz2-dev
-RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -q -y
+ARG RUST_VERSION=1.85.1
+
+# Install system dependencies
+RUN apt-get update && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    curl build-essential git upx-ucl libjemalloc-dev libjemalloc2 libbz2-dev && \
+    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -q -y --default-toolchain ${RUST_VERSION} && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+# Setup environment
+ENV PATH=$PATH:$HOME/.cargo/bin \
+    GOPROXY=direct \
+    GO111MODULE=on \
+    MAKEFLAGS="-j$(nproc)"
 
 WORKDIR /app
 
-# Copy source code
+# Copy dependency files first
+COPY go.mod go.sum ./
+COPY Makefile ./
+COPY vm/rust/Makefile vm/rust/
+COPY vm/rust/Cargo.toml vm/rust/Cargo.lock* vm/rust/
+COPY core/rust/Makefile core/rust/
+COPY core/rust/Cargo.toml core/rust/Cargo.lock* core/rust/
+COPY starknet/compiler/rust/Makefile starknet/compiler/rust/
+COPY starknet/compiler/rust/Cargo.toml starknet/compiler/rust/Cargo.lock* starknet/compiler/rust/
+
+# Setup dummy source files and fetch dependencies
+SHELL ["/bin/bash", "-c"]
+RUN mkdir -p vm/rust/src core/rust/src starknet/compiler/rust/src && \
+    touch vm/rust/src/lib.rs core/rust/src/lib.rs starknet/compiler/rust/src/lib.rs && \
+    go mod download && \
+    . $HOME/.cargo/env && \
+    cargo fetch --manifest-path vm/rust/Cargo.toml && \
+    cargo fetch --manifest-path core/rust/Cargo.toml && \
+    cargo fetch --manifest-path starknet/compiler/rust/Cargo.toml
```

```suggestion
    cargo build --manifest-path vm/rust/Cargo.toml && \
    cargo build --manifest-path core/rust/Cargo.toml && \
    cargo build --manifest-path starknet/compiler/rust/Cargo.toml
```
I think this will help caching the dependency build